### PR TITLE
wasm-bindgen-cli: 0.2.84 -> 0.2.87

### LINKED ISSUE #244276: Update request: wasm-bindgen-cli 0.2.84 → 0.2.87
- Package name:  wasm-bindgen-cli
- Latest released version: 0.28.4
- Current version on the unstable channel: 0.2.84
- Current version on the stable/release channel: 0.2.84

- [X] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

**Notify maintainers**
@nitsky and @Rizary

I'd really appreciate an update it's causing me problems with a trunk build similar to https://crane.dev/examples/trunk-workspace.html

-----

Note for maintainers: Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by jneem**
###### Description of changes

Fixes #244276
Closes #244334

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kirillrdy** [comment]
> Ok, there's a problem here that I don't know how to best resolve: since the last wasm-bindgen bump, lldap was added with a dependency on wasm-bindgen 0.2.84 in its Cargo.lock. Since the wasm-bindgen cli version needs to exactly match the crate version, this PR breaks the lldap build.
> 
> How can this be fixed? The only thing I can think of is to just package multiple versions of wasm-bindgen-cli.

if you look at lldaps lockfile you can see that nothing is actually requiring specific version of wasm-bindgen-cli, it will be much easier to just bump version of lldap lockfile instead of introducing multiple versions

File: pkgs/development/tools/wasm-bindgen-cli/default.nix [hunk 1]
```diff
@@ -8,27 +8,37 @@
 , curl
 , Security
 , runCommand
+, version ? "0.2.85"
 }:
 
+let hashes = {
+  "0.2.85" = {
```
Comment (kirillrdy) [inline pkgs/development/tools/wasm-bindgen-cli/default.nix:15]: 0.2.87 is out

File: pkgs/development/tools/wasm-bindgen-cli/default.nix [hunk 2]
```diff
@@ -8,27 +8,37 @@
...
   nativeBuildInputs = [ pkg-config ];
 
   buildInputs = [ openssl ] ++ lib.optionals stdenv.isDarwin [ curl Security ];
 
   nativeCheckInputs = [ nodejs ];
 
-  # other tests require it to be ran in the wasm-bindgen monorepo
-  cargoTestFlags = [ "--test=interface-types" ];
+  # tests require it to be ran in the wasm-bindgen monorepo
+  doCheck = false;
```
Comment (kirillrdy) [inline pkgs/development/tools/wasm-bindgen-cli/default.nix:41]: there is at least 1 passing test in 0.2.87 we should keep executing it

**Comment by jneem** [comment]
Hm, I think it would still be better to provide multiple wasm-bindgen versions, for two reasons:

- If we have only a single version in nixpkgs, we will need to patch the `Cargo.lock` of every rust package that uses `wasm-bindgen`. It would be much easier if we could just copy the upstream `Cargo.lock`s (which is currently the case for lldap, according to [this](https://github.com/NixOS/nixpkgs/blob/96f7ceb7b980489b79e9cc084bdb4b9481d421a6/pkgs/servers/ldap/lldap/default.nix#L52)) and then use whichever version of `wasm-bindgen-cli` is needed.
- having multiple versions is useful for devshells because then nixpkgs can be used to provide whichever version is needed

**Comment by kirillrdy** [comment]
> Hm, I think it would still be better to provide multiple wasm-bindgen versions, for two reasons:
> 
>     * If we have only a single version in nixpkgs, we will need to patch the `Cargo.lock` of every rust package that uses `wasm-bindgen`. It would be much easier if we could just copy the upstream `Cargo.lock`s (which is currently the case for lldap, according to [this](https://github.com/NixOS/nixpkgs/blob/96f7ceb7b980489b79e9cc084bdb4b9481d421a6/pkgs/servers/ldap/lldap/default.nix#L52)) and then use whichever version of `wasm-bindgen-cli` is needed.
> 
>     * having multiple versions is useful for devshells because then nixpkgs can be used to provide whichever version is needed

- usually multiple versions only added as last resort, eg true incompatibilities
- lldap doesn't depend on that specific version
- patching can coexist with Cargo.lock file copied from upstream repo ( eg `substituteAll` )  
- i don't object introducing multiple versions of wasm-bindgen-cli, it's all about balance of correctness and maintenance overhead 

in case for `lldap`, in my opinion there isn't enough to justify an extra version. If your main concern is ability to copy upstream Cargo.lock then its not an issue with `substituteAll` in `postPatch`

**Comment by jneem** [comment]
> If your main concern is ability to copy upstream Cargo.lock then its not an issue with `substituteAll` in `postPatch`

Just to check that I've understood correctly, are you suggesting having a `cargo-lock.patch` file containing some variables that get `substituteAll`ed out depending on the wasm-bindgen-cli version in nixpkgs?

I can give that a try, but at first glance it appears like more maintenance burden than keeping multiple wasm-bindgen-cli versions around. For example, lldap's `Cargo.lock` file appears to contain 5 crates that should be kept in sync (wasm-bindgen, wasm-bindgen-backend, wasm-bindgen-macro, wasm-bindgen-macro-support, and wasm-bindgen-shared). These all have the same version but different checksums, and so updating the variables to substitute would require looking up those 5 checksums on each wasm-bindgen-cli bump. Then the `cargo-lock.patch` would probably need to be recreated each time lldap gets updated because I wouldn't expect the old patch to apply to the new Cargo.lock. This is probably feasible as long as lldap is the only user of `wasm-bindgen-cli`, but I don't think it would scale very well.

In contrast, maintaining multiple versions of wasm-bindgen-cli just requires appending to the list of known hashes.

**Comment by kirillrdy** [comment]
> > If your main concern is ability to copy upstream Cargo.lock then its not an issue with `substituteAll` in `postPatch`
> 
> Just to check that I've understood correctly, are you suggesting having a `cargo-lock.patch` file containing some variables that get `substituteAll`ed out depending on the wasm-bindgen-cli version in nixpkgs?
> 
> I can give that a try, but at first glance it appears like more maintenance burden than keeping multiple wasm-bindgen-cli versions around. For example, lldap's `Cargo.lock` file appears to contain 5 crates that should be kept in sync (wasm-bindgen, wasm-bindgen-backend, wasm-bindgen-macro, wasm-bindgen-macro-support, and wasm-bindgen-shared). These all have the same version but different checksums, and so updating the variables to substitute would require looking up those 5 checksums on each wasm-bindgen-cli bump. Then the `cargo-lock.patch` would probably need to be recreated each time lldap gets updated because I wouldn't expect the old patch to apply to the new Cargo.lock. This is probably feasible as long as lldap is the only user of `wasm-bindgen-cli`, but I don't think it would scale very well.
> 
> In contrast, maintaining multiple versions of wasm-bindgen-cli just requires appending to the list of known hashes.


ok, I gave my approach a try and ... its a mess :-)  basically all wasm-bindgen-* crates need to be of a matching versions.

the only other advice I would have is, instead of having multiple `wasm-bindgen-cli` listed in `all-packages.nix,` only keep one, and  override `wasm-bindgen-cli` inside `lldap` derivation


motivation is, anything listed in all-packages.nix is top level attribute and it comes with certain expectations, eg when you remove things you need to list historical attributes in aliases.nix which will become very annoying,  especially since wasm-bindgen-cli get released fairly frequently

**Comment by jneem** [comment]
Thanks for checking so that I didn't have to :)

I've bumped to 0.2.87 and updated lldap to override the version. (Is there a good mechanism for storing multiple known hashes so that you only need to override the version and not the hashes?)

**Review by emilylange — APPROVED** [review]
The `(finalAttrs: {})`-pattern might make sense here, but shouldn't matter too much for now.

Also, thanks for looking after lldap! :)

I think locally pinning the version within `lldap/default.nix` via `.override` is a good idea.
And I hope it moves the maintenance burden from you to us (lldap maintainers).
I wasn't really aware how hard those version pins are.

lgtm!

**Pull request merged**